presto-spi/src/main/java/com/facebook/presto/spi/ConnectorSmallFragmentCoalescingPlan.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi;
+
+import io.airlift.slice.Slice;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Stream;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Objects.requireNonNull;
+
+public class ConnectorSmallFragmentCoalescingPlan
+{
+    private final Collection<Slice> retainedFragments;
```

Comment:
This structure worries me because there's a fundamental dependency between these three fields, but it's not enforced at all.  what if a fragment is neither in the retained nor deprecated lists?  What if it's in both?  What if it's in the retained list, but files inside it get coalesced?

Reply:
In the connector it's possible to check these invariants before returning the plan.

* Check that all original partitions are in the retained list (for statistics) and partition files are in either the retained or deprecated list
* Check that no file is in both